README.md

<p align="center">
  <a href="https://www.npmjs.com/package/calime" target="_blank" rel="noopener noreferrer">
    <img width="180" src="https://raw.githubusercontent.com/toolx-dev/calime/main/logo.png" alt="calime logo" />
  </a>
</p>
<br/>
<p align="center">
  <a href="https://www.npmjs.com/package/@toolx/core"><img src="https://img.shields.io/npm/v/calime" alt="npm package"></a>
  <img alt="NPM" src="https://img.shields.io/npm/l/calime">
</p>
<br/>

# calime

## Introduction
`Calime` is a versatile and lightweight class for styling and manipulating text in Node.js applications, particularly useful for console output. It offers a range of text styling options, such as color, background color, and text formatting styles like bold, underline, and italic. 

::: tip
While `Calime` provides various styling options, it aims to keep a balance between functionality and simplicity. For complex text and background color combinations, consider using libraries like `chalk` for more complete solutions. `Calime` may choose to skip adding such complex features to maintain code simplicity.
::::

 <img width="683" src="https://raw.githubusercontent.com/toolx-dev/calime/main/example.png" alt="calime example" />

## Features
- **Multiple Text Styles**: Apply various text styles including bold, underline, italic, and more.
- **Color Support**: Easily set text and background colors.
- **Text Gradients**: Create gradient effects on text using multiple colors.
- **Chainable Methods**: Methods can be chained for concise code.

## Installation

`calime` is available as an independent package on npm and can be installed in your Node.js project. Since it is exclusively an ESModule, ensure your environment supports ESModule syntax. To install `calime`, run the following command in your project directory:

```bash
npm install calime
```

## Usage
To use `Calime`, first import or include it in your Node.js project. Then, create a new instance of `Calime` and start styling your text.

```javascript
import Calime from 'calime'; // Adjust the import path as per your project structure

const myText = new Calime("Hello, World!");
myText.apply('bold').color([255, 0, 0]); // Applying bold style and red color
console.log(myText.render()); // Outputs styled text
```

## API Reference

### Constructor
- `Calime(text: string)` - Initialize with the specified text.

### Methods
- `apply(style: Style): Calime` - Apply a text style.
- `setText(text: string): Calime` - Set or update the text.
- `color(color: Color): Calime` - Set the text color.
- `gradient(...colors: Color[]): Calime` - Apply a gradient effect to the text.
- `background(color: Color): Calime` - Set the background color of the text.
- `render(): string` - Returns the styled text.

### Static Methods
- `static gradientColor(ratio: number, ...colors: Color[]): Color` - Compute gradient color.

## License
`Calime` is open-sourced MIT license and can be used freely in your projects.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/williammanco/calime/compare/v1.0.0...v1.0.1) (2023-12-29)

## 1.0.0 (2023-12-29)


test/index.test.js
const Calime = require('../src/index').default

describe('Calime', () => {
    let calime;

    beforeEach(() => {
        calime = new Calime("Initial Text");
    });

    test('constructor should initialize text correctly', () => {
        expect(calime.render()).toBe("Initial Text");
    });

    test('apply method should add style correctly', () => {
        calime.apply('bold');
        expect(calime.render()).toContain("\x1b[1m");
    });

    test('setText should update text correctly', () => {
        calime.setText("New Text");
        expect(calime.render()).toContain("New Text");
    });

    test('color should apply color correctly', () => {
        calime.color([255, 0, 0]); // Applying red color
        expect(calime.render()).toContain("\x1b[38;2;255;0;0m");
    });

    test('gradient should apply gradient correctly', () => {
        calime.gradient([255, 0, 0], [0, 255, 0]); // Gradient from red to green
        const renderedText = calime.render();
        expect(renderedText).toContain("\x1b[38;2;"); // Check for gradient color codes
    });

    test('background should apply background color correctly', () => {
        calime.background([0, 0, 255]); // Applying blue background
        expect(calime.render()).toContain("\x1b[48;2;0;0;255m");
    });

    test('static method gradientColor should compute colors correctly', () => {
        const color = Calime.gradientColor(0.5, [0, 0, 0], [255, 255, 255]); // Midpoint between black and white
        expect(color).toEqual([128, 128, 128]);
    });

    test('render should reset styles at the end', () => {
        calime.apply('bold');
        expect(calime.render()).toContain("\x1b[0m"); // Reset code at the end
    });
});


src/index.ts
type Style = 'reset' | 'bold' | 'underline' | 'dim' | 'inverse' | 'italic' | 'strikethrough';
type Color = [number, number, number];

class Calime {
    private text: string;
    private appliedStyles: string[];
    private appliedColor: Color | null;
    private appliedBackgroundColor: Color | null;
    private styles: Record<Style, string>;

    constructor(text: string) {
        this.text = text;
        this.appliedStyles = [];
        this.appliedColor = null;
        this.appliedBackgroundColor = null;
        this.styles = {
            reset: "\x1b[0m",
            bold: "\x1b[1m",
            underline: "\x1b[4m",
            dim: "\x1b[2m",
            inverse: "\x1b[7m",
            italic: "\x1b[3m",
            strikethrough: "\x1b[9m",
        };
    }

    apply(style: Style): this {
        if (!this.appliedStyles.includes(this.styles[style])) {
            this.appliedStyles.push(this.styles[style]);
        }
        return this;
    }

    setText(text: string): this {
        this.text = text;
        return this;
    }

    color(color: Color): this {
        this.appliedColor = color;
        return this;
    }

    gradient(...colors: Color[]): this {
        this.text = this.text.split('').map((char, index) => {
            const ratio = index / (this.text.length - 1);
            const gradientColor = Calime.gradientColor(ratio, ...colors);
            return `\x1b[38;2;${gradientColor[0]};${gradientColor[1]};${gradientColor[2]}m${char}`;
        }).join('');
        return this;
    }

    background(color: Color): this {
        this.appliedBackgroundColor = color;
        return this;
    }

    render(): string {
        let styleReset = '';
        if (this.appliedStyles.length > 0 || this.appliedColor || this.appliedBackgroundColor) {
            styleReset = this.styles.reset;
        }
        return `${this.appliedStyles.join('')}${this.appliedColor ? `\x1b[38;2;${this.appliedColor[0]};${this.appliedColor[1]};${this.appliedColor[2]}m` : ''}${this.appliedBackgroundColor ? `\x1b[48;2;${this.appliedBackgroundColor[0]};${this.appliedBackgroundColor[1]};${this.appliedBackgroundColor[2]}m` : ''}${this.text}${styleReset}`;
    }

    static gradientColor(ratio: number, ...colors: Color[]): Color {
        if (colors.length === 1) {
            return colors[0];
        }

        if (ratio < 0 || ratio > 1) {
            throw new Error("Ratio must be between 0 and 1.");
        }

        const step = 1 / (colors.length - 1);
        const index = Math.min(Math.floor(ratio / step), colors.length - 2);
        const colorStart = colors[index];
        const colorEnd = colors[index + 1];
        const localRatio = (ratio - index * step) / step;

        const r = Math.round(colorStart[0] + localRatio * (colorEnd[0] - colorStart[0]));
        const g = Math.round(colorStart[1] + localRatio * (colorEnd[1] - colorStart[1]));
        const b = Math.round(colorStart[2] + localRatio * (colorEnd[2] - colorStart[2]));

        return [r, g, b];
    }
}

export default Calime;
